### ISSUE TITLE
Add jupyter-ai (LLM) Support to JupyterLite

### ISSUE THREAD

**Issue by theonewolf** [jupyterlite/jupyterlite#1121]
### Problem

JupyterLab upstream has a new module to support LLMs while coding: `jupyter-ai`.  It would be great to take advantage of the latest LLM technology also inside JupyterLite.

### Proposed Solution

Similar to other upstream packages, test or create a compatibility package for `jupyter-ai` to integrate into JupyterLite.

### Additional context

This would help in using JupyterLite as a teaching tool, which is a huge killer application for it, because it requires no installation at all and runs entirely in a user's browser.

**Comment by theonewolf** [comment]
@jtpio I'm happy to work on this, but I do not know where to begin.  Do you have a wiki on how to make things compatible with JupyterLite?

**Comment by bollwyvl** [comment]
Please see the documentation for [emulating a server extension](https://jupyterlite.readthedocs.io/en/latest/howto/extensions/server.html) in a standalone extension.

This would likely _not_ be appropriate in jupyterlite core, as it requires interacting with a non-free SaaS which, at best, requires a for-pay, per-use token, which there would be no way to protect in a jupyterlite setting. At worst, and is all too common, the upstream service could break/go away.

**Comment by psychemedia** [comment]
A [recent PR](https://github.com/jupyterlab/jupyter-ai/pull/334) to `jupyter-ai` added support for the locally run LLMs as well as as hosted LLMs. If the locally hosted LLM can be accessed via an http API, then presumably there would be a route for the `jupyter-ai` extension to access it without any need to call on a non-free SaaS.

In the JupyterLite context, a WASM based LLM could also presumably be run in the browser context alongside JupyterLite, if your browser/machine was up to it? An early demo of WASM based LLM chat can be found here: https://github.com/mlc-ai/web-llm I'm not sure what models it supports, though?

**Comment by theonewolf** [comment]
@bollwyvl and @jtpio i was thinking we could take a pure JS approach with new WASM projects like Starcoder in JS:

https://rahuldshetty.github.io/starcoder.js/

I teach Python and Data Science.  Having a coding assistant directly available to my students in self-hosted notebooks would greatly enhance their learning experience.

@bollwyvl thanks for the extension documentation, I’ll take a look.

**Comment by jtpio** [comment]
FYI there is work happening in https://github.com/jupyterlite/ai.

Closing the issue since it's unlikely there will be anything to be done in the core JupyterLite repo. Don't hesitate to open new issues or discussions in https://github.com/jupyterlite/ai to discuss more, thanks!

**Comment by psychemedia** [comment]
BY the by, I note that AI models are starting to be exposed via js APIs. Summarizer API in [Chrome 138 Beta](https://developer.chrome.com/docs/ai/summarizer-api), other APIs in [Chrome 137 origin trials](https://developer.chrome.com/blog/ai-api-updates-io25).

**Comment by jtpio** [comment]
> BY the by, I note that AI models are starting to be exposed via js APIs. Summarizer API in [Chrome 138 Beta](https://developer.chrome.com/docs/ai/summarizer-api), other APIs in [Chrome 137 origin trials](https://developer.chrome.com/blog/ai-api-updates-io25).

Yeah, `window.LanguageModel` is already supported in https://github.com/jupyterlite/ai. You can use it on the GitHub Pages deployment (as long as the flags are enabled and other requirements met): https://jupyterlite.github.io/ai/

![Image](https://github.com/user-attachments/assets/757209ca-592d-4092-879f-d7d98bb557be)
